Bottom: 4a3b6b0cfc1e6e1685dd16ad0d554a91d6b2f5d3
Top:    139664cec39f2e531f468fbe9ca3f47885ad4291
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-29 16:45:07 -0500

Add codebase generator function

The codebase generator takes information about a change, like repository,
branch, and revision, and creates a predictable string describing the code.
This is used to decide upon the builder for a change and also to tie a fork
repository back to its parent for merge requests.

---
diff --git a/roles/buildbot-master/files/master.cfg b/roles/buildbot-master/files/master.cfg
index a12bdaa8..249bc9e3 100644
--- a/roles/buildbot-master/files/master.cfg
+++ b/roles/buildbot-master/files/master.cfg
@@ -6,6 +6,7 @@
 
 import os
 from buildbot.plugins import secrets, steps, util
+from buildbot.util import giturlparse
 
 os.environ['LONG_RUN_TIMEOUT'] = '14400'
 
@@ -142,3 +143,30 @@ c['workers'].extend(ptxdist.workers)
 c['schedulers'].extend(ptxdist.schedulers)
 c['builders'].extend(ptxdist.builders)
 c['change_source'].extend(ptxdist.change_source)
+
+
+def codebaseGenerator(chdict):
+    """ Generate a codebase description.
+
+    The codebase generator takes information about a change, like repository,
+    branch, and revision, and creates a predictable string describing the code.
+    This is used to decide upon the builder for a change and also to tie a fork
+    repository back to its parent for merge requests.
+    """
+
+    giturl = giturlparse(chdict['repository'])
+    return '/'.join((giturl.owner, giturl.repo))
+
+    # path = urlparse(chdict['revlink']).path.lstrip('/')
+
+    # components = []
+    # while path != "":
+    #     path, tail = os.path.split(path)
+    #     if tail != "":
+    #         components.append(tail)
+    # components.reverse()
+    # codebase = os.path.join(*components[0:2])
+    # return codebase
+
+
+c['codebaseGenerator'] = codebaseGenerator
